src/main.rs

```diff
@@ -145,6 +161,79 @@ type AFE1 = afe::ProgrammableGainAmplifier<
     hal::gpio::gpiod::PD15<hal::gpio::Output<hal::gpio::PushPull>>,
 >;
 
+/// Processes external timestamps to produce the frequency and initial phase of the demodulation
+/// signal.
+pub struct TimestampHandler {
+    pll: PLL,
+    pll_shift_frequency: u8,
+    pll_shift_phase: u8,
+    // Index of the current ADC batch.
+    batch_index: u32,
+    // Most recent phase and frequency values of the external reference.
+    reference_phase: i64,
+    reference_frequency: i64,
+}
+
+impl TimestampHandler {
+    /// Construct a new `TimestampHandler` instance.
+    ///
+    /// # Args
+    /// * `pll_shift_frequency` - See `PLL::update()`.
+    /// * `pll_shift_phase` - See `PLL::update()`.
+    ///
+    /// # Returns
+    /// New `TimestampHandler` instance.
+    pub fn new(pll_shift_frequency: u8, pll_shift_phase: u8) -> Self {
+        TimestampHandler {
+            pll: PLL::default(),
+            pll_shift_frequency,
+            pll_shift_phase,
+            batch_index: 0,
+            reference_phase: 0,
+            reference_frequency: 0,
+        }
+    }
+
+    /// Compute the initial phase and frequency of the demodulation signal.
+    ///
+    /// # Args
+    /// * `timestamp` - Counter value corresponding to an external reference edge.
+    ///
+    /// # Returns
+    /// Tuple consisting of the initial phase value and frequency of the demodulation signal.
+    pub fn update(&mut self, timestamp: Option<u32>) -> (u32, u32) {
```

Comment:
I'd set `ADC_SAMPLE_TICKS_LOG2` and `SAMPLE_BUFFER_SIZE_LOG2` as struct members (passed in `new()`) and move the struct and the impl into `dsp/src/reciprocal_pll.rs`.